### PR TITLE
Team IDK Proposal

### PR DESCRIPTION
**Pull request by Yanding-Liu**

### CODE REVIEW

File: teams/Team_IDK_description.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ 
+Placement of rooms:
+  Our large-scale map will be an x by y grid layout, with each room a subset of this x y grid. Once we’ve generated our room, we will randomly choose a location on our grid for the room to be placed (x1 - x2, y1 - y2). Before placing this room in this location, we will have to check if there already exists a room in that location or directly next to it. We can store this grid as a 2-D array containing a value that shows if it is taken. We will generate and place a certain number of rooms.
+	After placing all of our rooms on the map, we will then create a Delaunay triangulation graph from each of the rooms using the Bowyer-Watson algorithm. This will create connections between the rooms while keeping those connections relatively short. 
+ 
+Generating paths between rooms:
+  Once we’ve created the Delaunay triangulation graph, we will calculate the minimum spanning tree using Prim’s algorithm or Dijkstra’s algorithm to find the main routing path between the rooms. We will then randomly select a few other edges from the triangulation graph to create a few loops in the map layout.
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:63]: Dijkstra's is a shortest paths algorithm, not MST.

File: teams/Team_IDK_description.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ 
+   
+### ADVTOPIC2 - AI
+ 
+Our combat system would be a turn-based card game that takes place between the player and one enemy AI at a time. The main goal for the player is to reduce the enemy HP to 0 to win the battle, and vice versa. For the enemy AI, we decided to use reinforcement learning to train them through Q-learning. We think that for a turn-based game, Q-learning can suit the scenario well. It will update the state-action table through training. We will assign two terminal states to it, one is defeating the player, which will be rewarded greatly, and the other is being defeated, which is what we would like it to avoid. We would be assigning states based on the combination of actions made by the player and the current state of both sides. The algorithm will be based on the Bellman Equation for the value iteration update. The rewards that we assign will be based on damage dealt towards the player and damage taken by the AI. We might be training different sets of AIs by modifying the discount factor.
+
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:73]: To be clear, you're going to have to run the learning through a large number of simulated battles to train a policy that will actually be used in-game, correct?

File: teams/Team_IDK_description.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+## Concept Arts
+
+![Map](Team_IDK_Procedual_Generated_Layout.png)
+![Combat](Team_IDK_Turn-Based_Combat.JPG)
+
+## Midterm Goals
+ 
+* Have a working map/room
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:81]: I'm fine with a placeholder non-generated map for the midterm

File: teams/Team_IDK_description.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+* Player movement and collision 
+* Implement combat moveset/system
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:83]: How many moves to support? I'm fine with a naive dummy AI for the midterm

File: teams/Team_IDK_description.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+* Transition from exploration to combat
+* Have working UI
+ 
+ 
+## Final Goals
+ 
+* Functional turn-based combat (20%)
+  * Player can choose moves (10%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:91]: How many moves, how complex will the combat be?

File: teams/Team_IDK_description.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  * Transition between exploration screen and combat screen (5%)
+  * Display UI (5%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:93]: Specifically what UI elements do you want to display?

File: teams/Team_IDK_description.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+* Implement moveset for AI (30%)
+  * Implement Q-Learning Algorithm (10%)
+  * Have AI work with basic moveset (10%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:96]: Need more details on differences between basic and advanced movesets

File: teams/Team_IDK_description.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  * Have AI work with larger advanced moveset (10%)
+* Working procedural generated maps/rooms (30%)
+  * Randomly generated room (10%)
+  * Randomly generated map made of rooms (10%)
+  * Character can interact with map/room (10%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:101]: What iteractions specifically do you want to support?

File: teams/Team_IDK_description.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  * Player can choose moves (10%)
+  * Transition between exploration screen and combat screen (5%)
+  * Display UI (5%)
+* Implement moveset for AI (30%)
+  * Implement Q-Learning Algorithm (10%)
+  * Have AI work with basic moveset (10%)
+  * Have AI work with larger advanced moveset (10%)
+* Working procedural generated maps/rooms (30%)
+  * Randomly generated room (10%)
+  * Randomly generated map made of rooms (10%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:100]: How big of a map do you plan to generate?

File: teams/Team_IDK_description.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+* Functional turn-based combat (20%)
+  * Player can choose moves (10%)
+  * Transition between exploration screen and combat screen (5%)
+  * Display UI (5%)
+* Implement moveset for AI (30%)
+  * Implement Q-Learning Algorithm (10%)
+  * Have AI work with basic moveset (10%)
+  * Have AI work with larger advanced moveset (10%)
+* Working procedural generated maps/rooms (30%)
+  * Randomly generated room (10%)
```
Comment (nfarnan) [inline teams/Team_IDK_description.md:99]: How big will the rooms be? How many rooms per map?

**Review by nfarnan — CHANGES_REQUESTED** [review]

File: teams/Team_IDK_description.md (hunk 9 above)
Comment (mws69) [reply teams/Team_IDK_description.md:100]: 1000x1000 “tiles” within the 2-D array (grid)

File: teams/Team_IDK_description.md (hunk 8 above)
Comment (mws69) [reply teams/Team_IDK_description.md:101]: Character can interact with level exit, pick up keys, move around, collide with walls/objects

File: teams/Team_IDK_description.md (hunk 10 above)
Comment (mws69) [reply teams/Team_IDK_description.md:99]: 12 rooms per map, (ranging from 6x6 to 12x12 “tiles” within the 2-D array)

File: teams/Team_IDK_description.md (hunk 1 above)
Comment (mws69) [reply teams/Team_IDK_description.md:63]: *Prim's or Kruskal's

**Comment by nfarnan** [comment]
Please update the proposal and PR with the changes and updates mentioned in the discussion above.

File: teams/Team_IDK_description.md (hunk 5 above)
Comment (nfarnan) [reply teams/Team_IDK_description.md:94]: Given the updated description above, I'll require at least 8 moves

**Review by nfarnan — APPROVED** [review]

**Pull request merged**